content/blog/the-complete-beginners-guide-to-cli-apps-in-v.md

```diff
@@ -0,0 +1,837 @@
+
+Let us take a look at the files that were create by `v new`. In the folder, we
+find a `v.mod` file and a `geo.v` file.
+
+The `v.mod` file looks like this:
+
+```text {linenos=table}
+Module {
+	name: 'geo'
+	description: 'Sample CLI application that prints shapes to the screen.'
+	version: '0.0.1'
+	license: 'MIT'
+	dependencies: []
+}
+```
+
+And this is what `geo.v` contains:
+
+```v {linenos=table}
+module main
+
+fn main() {
+	println('Hello World!')
+}
+```
+
+Let us make sure that we can run the project out of the box by typing the
+command `v run .`:
+
+```text
+$ v run .
+Hello World!
+```
+
+## Accepting Command-Line Arguments
+
+Let us start with how we want to accept command-line arguments. We can make
+use of the in-built [`flag`](https://modules.vlang.io/flag.html) and
+[`os`](https://modules.vlang.io/os.html) modules to do this. In order to use
+these modules in our code, we will need to `import` them.
+
+We will start simple and put everything the `main` function, which is the
+entry point for the application. By default, all variables defined in V
+are immutable. To make a variable mutable, we need to use the `mut` keyword.
+We will make use of this when we define a new `FlagParser` struct using the
+`flag.new_flag_parser()` function.
+
+The `FlagParser` can then be modified to add details about the application.
+Additionally, we define the various types of flags that we want to accept
+using functions such as `bool(...)`, `int(...)`, `string(...)`, etc. The
+result of processing is returned as soon as we defined the flag which we
+store in various variables.
+
+Right now, we can just print out the values we've parsed to the screen.
+And when we're done, we need to call `finalize()` to finish the parsing,
+and obtain the rest of the arguments provided, if any.
+
+The final code that we will write is as follows:
+
+```v {linenos=table}
+module main
+
+import os
+import flag
+
+fn main() {
+	mut fp := flag.new_flag_parser(os.args)
+
+	fp.application('geo')
+	fp.version('0.0.1')
+	fp.description('A sample CLI application that prints geometric shapes to the console.')
+	fp.skip_executable()
+
+	shape := fp.string('shape', `p`, 'none', 'The shape to use for the geometry.').to_lower()
+
+	size := fp.int('size', `z`, 5, 'The size parameter for the shapes.')
+	symbol := fp.string('symbol', `m`, '*', 'The symbol to use for the geometry.').runes()[0] or {
+		`*`
+	}
+
+	additional_args := fp.finalize() ?
+
+	if additional_args.len > 0 {
+		println('Unprocessed arguments:\n$additional_args.join_lines()')
+	}
+
+	println(shape)
+	println(size)
+	println(symbol)
+}
+```
+
+We define our flag parse in line 7. We initialize it with details such as the
+name of the application, version, description, etc. Next, we define the various
+flags that we want to use. The general syntax is
+`` fp.<type>('<name>', `<short_name>`, <default_value>, '<description>') ``.
+
+Note that we use backticks to represent a single character, which is also known
```

Comment:
```suggestion
Note that we use backticks to represent a single UTF-8 character, which is also known
```

Reply:
Well, the file is stored with UTF-8 encoding, but you can store any Unicode character within backticks. I've tested it with emojis. The compound characters do not work though, like a gesture followed by skin tone; they need 2 runes. Probably better to not specify UTF-8 here because that would be irrelevant.

Reply:
If it isn't being used for a UTF-8 char, why not just use `byte`?  That's the only reason I added it - to give a reason to use this "special" type.

Reply:
No worries, that can be left as something for the user to explore.  :-)